Improvement GitHub Actions / Fix admin user existing

felddy opened the pull request:
This PR addresses the following issues:
- __Switch from Travis-CI to GitHub Actions__: The single Travis-CI configuration has been replaced with three GitHub Actions workflows.  One to lint and test pushes, one to handle prereleases, and one for full releases.   Breaking up the releases into separate workflows affords us more flexibility, and greatly simplifies the patcher script that injects parameters into the `packer.json` file.  Having separate workflows allows all variables to be adjusted, instead of relying on logic within the patcher script.  With this flexibility we will be able to easily deploy pre-releases to a staging account (playground), and full releases to production.
- __Fixes issue #6__: The debian AMI we are using comes with an `admin` user configured.  This user is used by Packer.  The user was persisting on to the final image with its ssh keys installed in both `admin` and `root`.  At the end of image creation, this user is now deleted, its sudo privileges revoked, and its keys are removed from `root`. (All keys are removed from `root`.)  Additionally, the cloud-init configuration is modified to prevent the user from being recreated at boot.
- __Re-enable ansible-linter__:  https://github.com/ansible/ansible-lint/pull/591 has been closed and merged, but not yet released.  The linter has been re-enabled in the pre-commit configuration, but is referenced by its commit SHA.

dav3r committed:
Add updates from running pre-commit autoupdate

dav3r committed:
Merge pull request #18 from cisagov/improvement/updates_from_precommit_autoupdate

Add updates from running pre-commit autoupdate

felddy committed:
Replace Travis-CI with GitHub actions.

felddy committed:
Fix action shield link

felddy committed:
Disable ansible-lint until they fix their setup.py
See: https://github.com/ansible/ansible-lint/issues/590

felddy committed:
Remove extra line.

felddy committed:
Merge pull request #20 from cisagov/improvement/actions

Replace Travis-CI with GitHub actions

felddy committed:
Merge branch 'develop' of github.com:cisagov/skeleton-generic into improvement/actions

felddy committed:
Convert travis config to workflow

felddy committed:
Add continuation characters

felddy committed:
Fix env var evaluations.

felddy committed:
Fix env variable calculation.

felddy committed:
Rename the iam user module.

felddy committed:
Add more environment variables

felddy committed:
Remove line that was driving Shane cray-cray.

felddy committed:
Add release event types to trigger

felddy committed:
Bump terraform version

felddy committed:
Break patcher into published and unpublished modes

felddy committed:
Remove github python library.
We are now have access to more state via actions

felddy committed:
Replace Travis job URL with GitHub commit SHA.

felddy committed:
Replace express double-quotes with signel-quotes.

felddy committed:
Work towards getting release tags out of GITHUB_REF

felddy committed:
Simplify patching of packer config.

felddy committed:
Simplify patching of packer config.

felddy committed:
Change environment variable name

felddy committed:
Remove boto3 requirement, and move semver.

felddy committed:
Split monolithic job into two workflows.

felddy committed:
Remove packer install from linter job

felddy committed:
Checkout the code we need to test.
Duh.

felddy committed:
Bump version.

felddy committed:
Add missing run command.

felddy committed:
Fix pre-release logic

felddy committed:
Add boto3 to requirements.
It is required for aws_ssm lookup.

felddy committed:
Add missing AWS variable.
It was hidden in plain sight.

felddy committed:
Change input from environment variable to stdin.

felddy committed:
Remove travis configuration.
So lange, süßer Prinz.

felddy committed:
Split unified release workflow into separate files.
This allows more customization based on the release type.  In the future 
we can use different AWS credentials putting the images in different 
accounts.  
This also removes branching logic that would run some steps sometimes.
It simplifies the environment variables, at the cost of having them 
spread in one more file.

felddy committed:
Move environment variables from the job to the workflow. 
See: 
https://github.blog/changelog/2019-10-01-github-actions-new-workflow-syntax-features/

felddy committed:
Remove unused environment variable.
This is checked during the build/test job and is not used during deploy.
The version during deploy is copied from the github tag.

felddy committed:
Update documentation to reflect changes to release workflows.

felddy committed:
Re-enable ansible-lint.
https://github.com/ansible/ansible-lint/issues/590 was closed.

felddy committed:
Auto-update pre-commit hooks.

felddy committed:
Change ansible-lint hook to point and fixed (unreleased) version.
See: 
https://github.com/ansible/ansible-lint/commit/7a2221a32968f12bee8494caf978ce150ad1494f

felddy committed:
Change job name to match workflow.

felddy committed:
Fix wrong environment variable name.

felddy committed:
Remove build account, and sudo access at end of build.
Addresses issue #6

felddy committed:
Attempt to at least clear the ssh authorized keys. 
The admin user was still on the AMI despite the shell provisioner.

T

felddy committed:
Expand command flags for clarity.

felddy committed:
Remove shell provisioner.
The user we delete is being recreated by cloud-init at boot.

felddy committed:
Change builder username so it more obvious on the server.

felddy committed:
User name must be admin.

felddy committed:
Re-add the shell provisioner with new cloud-init cleanup of default-user

felddy committed:
Add command to disable default user creation.
Replaces
users:
 - default

with:
users: []

felddy committed:
Replace last reference to Travis-CI.

felddy committed:
Sort workflow environment variables.

dav3r commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,61 @@
+                          us-west-1:alias/cool/ebs,
+                          us-west-2:alias/cool/ebs"
+  PACKER_VERSION: 1.4.4
+  TERRAFORM_VERSION: 0.12.12
```
It sure would be cool if we could find a nifty way to DRY out some of these environment variables that are the same in each of our workflow YML files... 😉

felddy replied:
I agree that would be nice.  I have looked.  There is no facility to do that at this time.  I expect that will change in the not-too-distant-future™.

felddy replied:
Just thought about this a bit more... We could do it using secrets.  It solves one problem but creates a pretty ugly other other one:  cannot see what versions we're using.   I hope they'll add support for external env files or an import/include syntax.

felddy replied:
We could also create a repo that holds our org-wide variables.  At runtime we could check it out and source it.   This could be a custom action.  Let me noodle on this some, and we'll revisit.

felddy replied:
The environment variables are now SUPER DRY.  Check out: f2fd977 and https://github.com/cisagov/setup-env-github-action

dav3r commented on .pre-commit-config.yaml:
```diff
@@ -67,7 +67,7 @@ repos:
       - id: black
   - repo: https://github.com/ansible/ansible-lint.git
-    rev: v4.1.1a0
+    rev: 7a2221a32968f12bee8494caf978ce150ad1494f
```
Can you open an issue to remind us to change this back to a release (when one becomes available)?

felddy replied:
I'm going to backport this change up to skeleton-generic.  I'll modify the issue that is up there: https://github.com/cisagov/skeleton-generic/issues/19

dav3r commented on README.md:
```diff
@@ -1,98 +1,101 @@
 If you are satisfied with your pre-release image, you can easily create a release
-that deploys to all regions by changing the `pre-release` command of
-`patch_packer_config.py` to `release` and rerunning packer:
+that deploys to all regions by changing the input to
```
Are you talking about deploying a pre-release to all regions?  Or are you talking about a real release here?  This section is a little confusing because your examples don't show the value of the `GITHUB_IS_PRERELEASE` env variable.

felddy replied:
This section is not about using GitHub to do true pre-releases, and releases.  This is hand-jamming it on your local box (for @jsf9k) .  The `GITHUB_IS_PRERELEASE` variable is optional and only used to tag the AMI.  

Make sure to take a look at the `README` as a whole, and not just the diff.  I think it is a bit clearer when you see the final doc.

mcdonnnj replied:
I recommend switching to the side-by-side diff view to more easily compare the two documents as wholes instead of intermingled.

dav3r reviewed (approved):
This is looking swell- great work here @felddy.  Please review my comments though before merging.
⚙️ ⚙️ ⚙️

felddy committed:
Use new github action to set our tool versions via env. 
DRYs out the code a bit, and removes the need to bump tools versions in 
all repos.

mcdonnnj committed:
Some small punctuation changes to help readability.

mcdonnnj reviewed (commented):
This is some nice work. I like the transition to GitHub Actions and this is a good transition PR. Strong work @felddy 💪

jsf9k commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,60 @@
+            --time-cond "${CURL_CACHE_DIR}/${PACKER_ZIP}" --location \
+            "https://releases.hashicorp.com/packer/${PACKER_VERSION}/${PACKER_ZIP}"
+          sudo unzip -o -d /usr/local/bin "${CURL_CACHE_DIR}/${PACKER_ZIP}"
+      - name: Install Terraform
```
Could the install of packer and terraform be broken out into a separate repo a la what you did with [cisagov/setup-env-github-action](https://github.com/cisagov/setup-env-github-action)?

Seems maybe the Ansible role stuff could possibly be broken out as well, taking the list of role names as an input.

jsf9k reviewed (commented):
This is strong work.  I left a comment discussing how things could possibly be improved (DRYed) even further.
